iceoryx2-ffi/cxx/include/iox2/sample.hpp

```diff
@@ -13,59 +13,137 @@
+    auto payload() const -> const Payload&;
 
-    auto payload() const -> const Payload& {
-        IOX_TODO();
-    }
-    auto header() const -> const HeaderPublishSubscribe& {
-        IOX_TODO();
+    /// Returns a reference to the user_header of the [`Sample`]
+    template <typename T = UserHeader, typename = std::enable_if_t<!std::is_same_v<void, UserHeader>, T>>
+    auto user_header() const -> const T&;
+
+    /// Returns a reference to the [`Header`] of the [`Sample`].
+    auto header() const -> const HeaderPublishSubscribe&;
+
+    /// Returns the [`UniquePublisherId`] of the [`Publisher`](crate::port::publisher::Publisher)
+    auto origin() const -> UniquePublisherId;
+
+  private:
+    template <ServiceType, typename, typename>
+    friend class Subscriber;
+
+    explicit Sample(iox2_sample_h handle);
+    void drop();
+
+    iox2_sample_h m_handle { nullptr };
+};
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline Sample<S, Payload, UserHeader>::Sample(iox2_sample_h handle)
+    : m_handle { handle } {
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline void Sample<S, Payload, UserHeader>::drop() {
+    if (m_handle != nullptr) {
+        iox2_sample_drop(m_handle);
+        m_handle = nullptr;
     }
-    auto origin() const -> UniquePublisherId {
-        IOX_TODO();
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline Sample<S, Payload, UserHeader>::Sample(Sample&& rhs) noexcept {
+    *this = std::move(rhs);
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline auto Sample<S, Payload, UserHeader>::operator=(Sample&& rhs) noexcept -> Sample& {
+    if (this != &rhs) {
+        drop();
+        m_handle = std::move(rhs.m_handle);
+        rhs.m_handle = nullptr;
     }
-};
+
+    return *this;
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline Sample<S, Payload, UserHeader>::~Sample() {
+    drop();
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline auto Sample<S, Payload, UserHeader>::operator*() const -> const Payload& {
+    return payload();
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline auto Sample<S, Payload, UserHeader>::operator->() const -> const Payload* {
+    return &payload();
+}
+
+template <ServiceType S, typename Payload, typename UserHeader>
+inline auto Sample<S, Payload, UserHeader>::payload() const -> const Payload& {
+    auto* ref_handle = iox2_cast_sample_ref_h(m_handle);
+    const Payload* payload_ptr = nullptr;
+    size_t payload_len = 0;
+
+    // NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast), no other way for type erasure
+    iox2_sample_payload(ref_handle, reinterpret_cast<const void**>(&payload_ptr), &payload_len);
+    // NOLINTEND(cppcoreguidelines-pro-type-reinterpret-cast)
+    IOX_ASSERT(sizeof(Payload) <= payload_len, "");
+
+    return *payload_ptr;
```

Comment:
If you want to get rid of the `reinterpret_cast`, this is an option
```suggestion
    auto* ref_handle = iox2_cast_sample_ref_h(m_handle);
    const void* payload_ptr = nullptr;
    size_t payload_len = 0;

    iox2_sample_payload(ref_handle, &payload_ptr, &payload_len);
    IOX_ASSERT(sizeof(Payload) <= payload_len, "");

    return *static_cast<const Payload*>(payload_ptr);
```